hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx

```diff
@@ -343,41 +152,298 @@ let cancelSignal: AbortController;
   autoReload: AutoReloadHelper;
+  columns: ColumnProps<IDatanode>[];
+  defaultColumns: IOption[];
 
   constructor(props = {}) {
     super(props);
+    this.columns = [
+      {
+        title: 'Hostname',
+        dataIndex: 'hostname',
+        key: 'hostname',
+        isVisible: true,
+        isSearchable: true,
+        sorter: (a: IDatanode, b: IDatanode) => a.hostname.localeCompare(b.hostname, undefined, {numeric: true}),
+        defaultSortOrder: 'ascend' as const,
+        fixed: 'left'
+      },
+      {
+        title: 'State',
+        dataIndex: 'state',
+        key: 'state',
+        isVisible: true,
+        isSearchable: true,
+        filterMultiple: true,
+        filters: DatanodeStateList && DatanodeStateList.map(state => ({text: state, value: state})),
+        onFilter: (value: DatanodeState, record: IDatanode) => record.state === value,
+        render: (text: DatanodeState) => renderDatanodeState(text),
+        sorter: (a: IDatanode, b: IDatanode) => a.state.localeCompare(b.state)
+      },
+      {
+        title: 'Operational State',
+        dataIndex: 'opState',
+        key: 'opState',
+        isVisible: true,
+        isSearchable: true,
+        filterMultiple: true,
+        filters: DatanodeOpStateList && DatanodeOpStateList.map(state => ({text: state, value: state})),
+        onFilter: (value: DatanodeOpState, record: IDatanode) => record.opState === value,
+        render: (text: DatanodeOpState) => renderDatanodeOpState(text),
+        sorter: (a: IDatanode, b: IDatanode) => a.opState.localeCompare(b.opState)
+      },
+      {
+        title: 'Uuid',
+        dataIndex: 'uuid',
+        key: 'uuid',
+        isVisible: true,
+        isSearchable: true,
+        sorter: (a: IDatanode, b: IDatanode) => a.uuid.localeCompare(b.uuid),
+        defaultSortOrder: 'ascend' as const
+      },
+      {
+        title: 'Storage Capacity',
+        dataIndex: 'storageCapacity',
+        key: 'storageCapacity',
+        isVisible: true,
+        filters: [
+          {
+            text: 'Storage Remaining (default)',
+            value: 'storageRemaining',
+          },
+          {
+            text: 'Storage Used',
+            value: 'storageUsed',
+          },
+          {
+            text: 'Storage Committed',
+            value: 'storageCommitted'
+          },
+          {
+            text: 'Storage Total',
+            value: 'storageTotal',
+          },
+          {
+            text: 'Storage Utilization',
+            value: 'storageUtilization',
+          }
+        ],
+        filterIcon: () => (
+          <Icon type={'sort-ascending'} />
+        ),
+        filterMultiple: false,
+        onFilter: () => {
+          // This filter is only used for placeholder, so return true regardless of the filter
+          return true;
+        },
+        sorter: (a: IDatanode, b: IDatanode) => {
+          let sortBy = "storageRemaining"
+          if (this.state.filteredInfo.storageCapacity &&
```

Comment:
Not clear with this part, whenever user tries to sort, here we are checking storageCapacity column length and has data, and if yes, then we are setting `sortBy=first filter value ` always  ? Am i missing something here ?

Reply:
Thank you for the review.

> here we are checking storageCapacity column length and has data, and if yes, then we are setting sortBy=first filter value  always ?

From the AntD table API documentation (https://3x.ant.design/components/table/#API). 

<img width="1026" alt="image" src="https://github.com/apache/ozone/assets/36403683/d9510c7c-165f-4211-9637-705ee90862a7">

The filter value has a type `string[]` and therefore the filtered info has the type `Partial<Record<string, string[]>>` (in `IDatanodeState`).

So when we change the filter value, the filtered info would be something like

Storage remaining
```
{
  "storageCapacity": ["storageRemaining"] 
}
```

Storage Used
```
{
  "storageCapacity": ["storageUsed"] 
}

The reason why the filter only contains 1 element is that `filterMultiple` for the 'Storage Capacity' column is set to `false`. If it's set to true, user can pick more than one filter and the array will be larger than 1. 

Therefore, since at any point of time, at most one filter value can be picked, we can pick the first value of the array after doing the nullity check and length check. If the filter is not picked yet, we will pick the default storage metric (in this case `storageRemaining`)

Hopefully this clarifies.